src/garage/torch/modules/gaussian_mlp_module.py

```diff
@@ -0,0 +1,255 @@
+                 learn_std=True,
+                 init_std=1.0,
+                 min_std=1e-6,
+                 max_std=None,
+                 std_hidden_sizes=(32, 32),
+                 std_hidden_nonlinearity=torch.tanh,
+                 std_hidden_w_init=nn.init.xavier_uniform_,
+                 std_hidden_b_init=nn.init.zeros_,
+                 std_output_nonlinearity=None,
+                 std_output_w_init=nn.init.xavier_uniform_,
+                 std_parameterization='exp',
+                 layer_normalization=False):
+        super(GaussianMLPBaseModule, self).__init__()
+
+        self._input_dim = input_dim
+        self._hidden_sizes = hidden_sizes
+        self._action_dim = output_dim
+        self._learn_std = learn_std
+        self._std_hidden_sizes = std_hidden_sizes
+        self._min_std = min_std
+        self._max_std = max_std
+        self._std_hidden_nonlinearity = std_hidden_nonlinearity
+        self._std_hidden_w_init = std_hidden_w_init
+        self._std_hidden_b_init = std_hidden_b_init
+        self._std_output_nonlinearity = std_output_nonlinearity
+        self._std_output_w_init = std_output_w_init
+        self._std_parameterization = std_parameterization
+        self._hidden_nonlinearity = hidden_nonlinearity
+        self._hidden_w_init = hidden_w_init
+        self._hidden_b_init = hidden_b_init
+        self._output_nonlinearity = output_nonlinearity
+        self._output_w_init = output_w_init
+        self._output_b_init = output_b_init
+        self._layer_normalization = layer_normalization
+
+        if self._std_parameterization not in ('exp', 'softplus'):
+            raise NotImplementedError
+
+        self._init_std_param = torch.Tensor([init_std])
+        if self._learn_std:
+            self._init_std = torch.nn.Parameter(self._init_std_param)
+
+        self._init_std_param = self._init_std_param.log()
+
+        self._min_std_param = self._max_std_param = None
+        if min_std is not None:
+            self._min_std_param = torch.tensor(min_std, dtype=torch.float).log()
+        if max_std is not None:
+            self._max_std_param = torch.tensor(max_std, dtype=torch.float).log()
+
+    @abc.abstractmethod
+    def _get_mean_and_log_std(self, inputs):
+        pass
+
+    def forward(self, inputs):
+        mean, log_std_uncentered = self._get_mean_and_log_std(inputs)
+
+        if self._min_std_param or self._max_std_param:
+            log_std_uncentered = log_std_uncentered.clamp(
+                min=self.to_scalar_if_not_none(self._min_std_param),
+                max=self.to_scalar_if_not_none(self._max_std_param))
+
+        if self._std_parameterization == 'exp':
+            std = log_std_uncentered.exp()
+        else:
+            std = log_std_uncentered.exp().exp().add(1.).log()
+
+        cov = (std**2).diag_embed()
+        dist = MultivariateNormal(mean, cov)
+
+        return dist
+
+    def to_scalar_if_not_none(self, tensor):
```

Comment:
@krzentner WDTY of this?

Reply:
I would just make it a private function of the module. Otherwise I'm fine with it.

It would definitely be a code smell in Java, but using small intermediate functions is relatively normal in Python.

Reply:
fixed